File: frontend/src/sdk/counter.ts
import { api } from './api'

export const StatusStopped = 0
export const StatusPause = 1
export const StatusRunning = 2

export type Status = typeof StatusStopped | typeof StatusPause | typeof StatusRunning

export type Counter = {
	name: string
	init: number
	count: number
	status: Status
}

export function GetAll() {
	return api<Record<string, Counter>>('/counter', 'GET')
}

export function Get(name: string) {
	return api<Counter>(`/counter/${name}`, 'GET')
}

export function SetInit(name: string, init: number) {
	return api<Counter>(`/counter/${name}`, 'PUT', {
		init,
	})
}

export function Start(name: string) {
	return api<Counter>(`/counter/${name}/start`, 'PUT')
}

export function Stop(name: string) {
	return api<Counter>(`/counter/${name}/stop`, 'PUT')
}

export function Reset(name: string) {
	return api<Counter>(`/counter/${name}/reset`, 'PUT')
}

File: backend/api/counter.go
package api

import (
	"context"
	"net/http"

	"backend/models/counter"

	"github.com/danielgtaylor/huma/v2"
)

func (h *Handler) GetAllCounter(ctx context.Context, input *struct{}) (*Output[counter.CounterGroup], error) {
	return &Output[counter.CounterGroup]{
		Body: h.counters,
	}, nil
}

func (h *Handler) GetCounter(ctx context.Context, input *struct {
	Name string `path:"name" doc:"counter name" example:"R0"`
},
) (*Output[counter.Counter], error) {
	c := h.counters.Get(input.Name)
	if c == nil {
		return nil, huma.NewError(http.StatusNotFound, "counter not found")
	}

	return &Output[counter.Counter]{
		Body: *c,
	}, nil
}

type CreateInit struct {
	Init int `json:"init" doc:"initial value" example:"10"`
}

func (h *Handler) SetCounterInit(ctx context.Context, input *struct {
	Name string     `path:"name" doc:"counter name" example:"R0"`
	Body CreateInit `json:"body" doc:"counter init value"`
},
) (*Output[counter.Counter], error) {
	c := h.counters.Get(input.Name)
	if c == nil {
		return nil, huma.NewError(http.StatusNotFound, "counter not found")
	}

	c.SetInit(input.Body.Init)

	return &Output[counter.Counter]{
		Body: *c,
	}, nil
}

func (h *Handler) CounterStart(ctx context.Context, input *struct {
	Name string `path:"name" doc:"counter name" example:"R0"`
},
) (*Output[counter.Counter], error) {
	c := h.counters.Get(input.Name)
	if c == nil {
		return nil, huma.NewError(http.StatusNotFound, "counter not found")
	}

	go c.Start()

	return &Output[counter.Counter]{
		Body: *c,
	}, nil
}

func (h *Handler) CounterStop(ctx context.Context, input *struct {
	Name string `path:"name" doc:"counter name" example:"R0"`
},
) (*Output[counter.Counter], error) {
	c := h.counters.Get(input.Name)
	if c == nil {
		return nil, huma.NewError(http.StatusNotFound, "counter not found")
	}

	c.Stop()

	return &Output[counter.Counter]{
		Body: *c,
	}, nil
}

func (h *Handler) CounterReset(ctx context.Context, input *struct {
	Name string `path:"name" doc:"counter name" example:"R0"`
},
) (*Output[counter.Counter], error) {
	c := h.counters.Get(input.Name)
	if c == nil {
		return nil, huma.NewError(http.StatusNotFound, "counter not found")
	}

	c.Reset()

	return &Output[counter.Counter]{
		Body: *c,
	}, nil
}
